services/hanke-service/src/main/kotlin/fi/hel/haitaton/hanke/hakemus/HakemusUpdateRequest.kt

```diff
@@ -0,0 +1,146 @@
+    override val customerWithContacts: HakemusUpdateRequestCustomerWithContacts? = null,
+    /** Työn suorittajan tiedot */
+    val contractorWithContacts: HakemusUpdateRequestCustomerWithContacts? = null,
+    /** Rakennuttajan tiedot */
+    val propertyDeveloperWithContacts: HakemusUpdateRequestCustomerWithContacts? = null,
+    /** Asianhoitajan tiedot */
+    override val representativeWithContacts: HakemusUpdateRequestCustomerWithContacts? = null,
+    // 4. sivu Liitteet (separete endpoint)
+    // 5. sivu Yhteenveto (no input data)
+) : HakemusUpdateRequest {
+
+    /**
+     * Returns true if this application data has changes compared to the given [applicationData].
+     */
+    fun hasChanges(applicationData: CableReportApplicationData): Boolean =
+        name != applicationData.name ||
+            streetAddress != applicationData.postalAddress?.streetAddress?.streetName ||
+            constructionWork != applicationData.constructionWork ||
+            maintenanceWork != applicationData.maintenanceWork ||
+            propertyConnectivity != applicationData.propertyConnectivity ||
+            emergencyWork != applicationData.emergencyWork ||
+            rockExcavation != applicationData.rockExcavation ||
+            workDescription != applicationData.workDescription ||
+            startTime != applicationData.startTime ||
+            endTime != applicationData.endTime ||
+            areas != applicationData.areas ||
+            customerWithContacts?.hasChanges(applicationData.customerWithContacts)
+                ?: !applicationData.customerWithContacts.isNullOrBlank() ||
+            contractorWithContacts?.hasChanges(applicationData.contractorWithContacts)
+                ?: !applicationData.contractorWithContacts.isNullOrBlank() ||
+            propertyDeveloperWithContacts?.hasChanges(applicationData.propertyDeveloperWithContacts)
+                ?: !applicationData.propertyDeveloperWithContacts.isNullOrBlank() ||
+            representativeWithContacts?.hasChanges(applicationData.representativeWithContacts)
+                ?: !applicationData.representativeWithContacts.isNullOrBlank()
+}
+
+data class HakemusUpdateRequestCustomerWithContacts(
+    val customer: HakemusUpdateRequestCustomer,
+    val contacts: List<HakemusUpdateRequestContact>
+) {
+    /**
+     * Returns true if this customer has changes compared to the given [customerWithContacts]. For
+     * contacts, we check that both have the same ids (order doesn't matter).
+     */
+    fun hasChanges(customerWithContacts: CustomerWithContacts?): Boolean =
+        customerWithContacts?.let {
+            customer.hasChanges(it.customer) ||
+                contacts.size != customerWithContacts.contacts.size ||
+                contacts.map { contact -> contact.id }.toSet() !=
+                    customerWithContacts.contacts.map { contact -> contact.id }.toSet()
+        } ?: !customer.isEmpty()
+}
+
+data class HakemusUpdateRequestCustomer(
```

Comment:
This will want an `id` field, so we know if the user wants to update an existing customer or create a new one.

Reply:
Ah, yes! Added the id. Thanks!